File: problems/array-string/medium/006-zigzag-conversion/Solution.kt
class Solution {
	fun convert(s: String, numRows: Int): String {
		if (numRows == 1 || s.length <= numRows) {
			return s
		}

		val zigzag = MutableList(numRows) { StringBuilder() }
		var isBackward = false
		var rowIndex = 0

		for (ch in s) {
			zigzag[rowIndex].append(ch)

			if (isBackward) {
				rowIndex--
			} else {
				rowIndex++
			}

			if (rowIndex == 0) {
				isBackward = false
			} else if (rowIndex == numRows - 1) {
				isBackward = true
			}
		}

		val result = StringBuilder()
		for (row in zigzag) {
			result.append(row)
		}
		return result.toString()
	}
}

File: problems/array-string/medium/006-zigzag-conversion/Solution.cpp
#include <string>
#include <vector>
using namespace std;

class Solution
{
public:
    string convert(string s, int numRows)
    {
        if (numRows == 1)
        {
            return s;
        }
        vector<string> zigzag(numRows, "");
        bool isBackward = false;
        int rowIndex = 0;
        for (int i = 0; i < s.length(); i++)
        {
            zigzag[rowIndex] += s[i];
            if (isBackward)
            {
                rowIndex--;
            }
            else
            {
                rowIndex++;
            }
            if (rowIndex == 0)
            {
                isBackward = false;
            }
            else if (rowIndex == numRows - 1)
            {
                isBackward = true;
            }
        }
        string result;
        for (const string &row : zigzag)
        {
            result += row;
        }
        return result;
    }
};
